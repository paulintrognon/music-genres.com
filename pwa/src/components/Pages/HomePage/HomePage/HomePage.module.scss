.container {
  padding-top: 5rem;
}

.title {
  font-family: BebasNeue;
  text-transform: uppercase;
  font-size: var(--font-size-30);
  text-align: center;
}

.subtitle {
  text-align: center;
  font-size: var(--font-size-18);
  margin-top: 0.5rem;
}
