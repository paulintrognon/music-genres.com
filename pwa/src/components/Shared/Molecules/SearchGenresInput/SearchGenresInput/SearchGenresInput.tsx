import { ChangeEvent, useState } from 'react'
import useSWR from 'swr'
import { GenreSuggestionsApiResultType } from '../../../../../types/Api/GenreSuggestionsApiResultType'
import MagnifyingGlassIcon from '../../../Atoms/Icons/MagnifyingGlass/MagnifyingGlass'
import InputWithIcon from '../../InputWithIcon/InputWithIcon'
import GenresSuggestions from '../GenresSuggestions/GenresSuggestions'
import classes from './SearchGenresInput.module.scss'

interface Props {
  slugToHref: (slug: string) => string
  onChange?: (entry: string) => void
  onSubmit?: (entry: string) => void
  onDirty?: () => void
  onClean?: () => void
}
const SearchGenresInput: React.FC<Props> = ({
  slugToHref,
  onChange,
  onSubmit,
  onDirty,
  onClean,
}) => {
  /**
   * On string query changes, we update query state
   */
  const [typedText, setTypedText] = useState('')
  const handleOnChange = (event: ChangeEvent<HTMLInputElement>): void => {
    setTypedText(event.target.value)
    onChange && onChange(event.target.value)
  }

  /**
   * On query state changes, we fetch up to 5 genres suggestions
   */
  const { data, isValidating } = useSWR<GenreSuggestionsApiResultType>(
    typedText && `/api/music-genres/search?query=${typedText}&limit=5`
  )

  /**
   * On submit, we pass the typed text to onSubmit handler (if exists)
   */
  const handleOnSubmit = (): void => {
    onSubmit && onSubmit(typedText)
  }

  return (
    <div className={classes.container}>
      <InputWithIcon
        className={classes.searchInput}
        placeholder="Start looking for a genre"
        icon={<MagnifyingGlassIcon />}
        onDirty={onDirty}
        onClean={onClean}
        onChange={handleOnChange}
        onSubmit={handleOnSubmit}
      />
      {typedText ? (
        <GenresSuggestions
          isLoading={isValidating}
          genreName={typedText}
          suggestions={data?.result || []}
          slugToHref={slugToHref}
        />
      ) : null}
    </div>
  )
}
export default SearchGenresInput
