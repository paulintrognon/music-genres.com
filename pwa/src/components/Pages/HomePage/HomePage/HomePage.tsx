import { useState } from 'react'
import { genreLink } from '../../../../services/links/links'
import SearchGenresInput from '../../../Shared/Molecules/SearchGenresInput/SearchGenresInput/SearchGenresInput'
import GenresExamples from '../GenresExamples/GenresExamples'
import classes from './HomePage.module.scss'

const HomePage: React.FC = () => {
  /**
   * When the user focuses the search input, we hide the elements bellow.
   * On blur, we show them again.
   */
  const [bellowSearchElementsOpacity, setBellowSearchElementsOpacity] = useState(1)
  const handleOnChangeSearchFocus = (focus: boolean) => () => {
    focus ? setBellowSearchElementsOpacity(0) : setBellowSearchElementsOpacity(1)
  }

  return (
    <section className={classes.container}>
      <h2 className={classes.title}>Discover genres by listening to what they sound</h2>
      <p className={classes.subtitle}>
        Music Genres gives you a sonore definition of musical genres by listening to YouTube videos.
      </p>
      <SearchGenresInput
        slugToHref={genreLink}
        onDirty={handleOnChangeSearchFocus(true)}
        onClean={handleOnChangeSearchFocus(false)}
      />
      <div
        data-cy="HomePage_BellowSearchElements"
        style={{ opacity: bellowSearchElementsOpacity, transition: 'opacity 200ms' }}
      >
        <GenresExamples />
      </div>
    </section>
  )
}
export default HomePage
