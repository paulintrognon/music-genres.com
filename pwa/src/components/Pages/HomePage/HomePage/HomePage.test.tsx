import { shallow } from 'enzyme'
import HomePage from './HomePage'
import SearchGenresInput from '../../../Shared/Molecules/SearchGenresInput/SearchGenresInput/SearchGenresInput'

it('should render the HomePage', () => {
  const wrapper = shallow(<HomePage />)
  expect(wrapper).toMatchSnapshot()
})

it('should change the opacity of components bellow search input', () => {
  const wrapper = shallow(<HomePage />)

  // Initially opacity should be 1
  expect(wrapper.find('[data-cy="HomePage_BellowSearchElements"]').props().style).toHaveProperty(
    'opacity',
    1
  )

  // On search focus, opacity should be 0
  ;(wrapper.find(SearchGenresInput).props().onDirty as () => void)()
  expect(wrapper.find('[data-cy="HomePage_BellowSearchElements"]').props().style).toHaveProperty(
    'opacity',
    0
  )

  // On search blur, opacity should go back to 1
  ;(wrapper.find(SearchGenresInput).props().onClean as () => void)()
  expect(wrapper.find('[data-cy="HomePage_BellowSearchElements"]').props().style).toHaveProperty(
    'opacity',
    1
  )
})
