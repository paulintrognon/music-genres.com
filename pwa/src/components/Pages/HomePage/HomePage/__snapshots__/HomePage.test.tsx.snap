// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the HomePage 1`] = `
<section
  className="container"
>
  <h2
    className="title"
  >
    Discover genres by listening to what they sound
  </h2>
  <p
    className="subtitle"
  >
    Music Genres gives you a sonore definition of musical genres by listening to YouTube videos.
  </p>
  <SearchGenresInput
    onClean={[Function]}
    onDirty={[Function]}
    slugToHref={[Function]}
  />
  <div
    data-cy="HomePage_BellowSearchElements"
    style={
      Object {
        "opacity": 1,
        "transition": "opacity 200ms",
      }
    }
  >
    <GenresExamples />
  </div>
</section>
`;
